Port files trashbin events to IEventDispatcher/IEventListener

CarlSchwan opened the pull request:
oc_hooks is deprecated and will at some point be removed

CarlSchwan committed:
feat(trashbin): Port files trashbin events to IEventDispatcher/IEventListener

oc_hooks is deprecated and will at some point be removed

Signed-off-by: Carl Schwan <carl@carlschwan.eu>
Signed-off-by: Thomas Citharel <tcit@tcit.fr>
Signed-off-by: Côme Chilliet <come.chilliet@nextcloud.com>

artonge reviewed (commented):
Still a bunch of lint errors/warnings :)

come-nc commented on apps/files_trashbin/lib/AppInfo/Application.php:
```diff
@@ -55,20 +62,18 @@ public function register(IRegistrationContext $context): void {
 		$context->registerServiceAlias('principalBackend', Principal::class);
 
 		$context->registerUserMigrator(TrashbinMigrator::class);
+
+		$context->registerEventListener(NodeWrittenEvent::class, EventListener::class);
+		$context->registerEventListener(BeforeUserDeletedEvent::class, EventListener::class);
+
+		// pre and post-rename, disable trash logic for the copy+unlink case
+		$context->registerEventListener(BeforeNodeDeletedEvent::class, Trashbin::class);
+		$context->registerEventListener(BeforeFileSystemSetupEvent::class, EventListener::class);
```
```suggestion

		$context->registerEventListener(NodeWrittenEvent::class, EventListener::class);
		$context->registerEventListener(BeforeUserDeletedEvent::class, EventListener::class);
		$context->registerEventListener(BeforeFileSystemSetupEvent::class, EventListener::class);

		// pre and post-rename, disable trash logic for the copy+unlink case
		$context->registerEventListener(BeforeNodeDeletedEvent::class, Trashbin::class);
```
Please reorder to make it clearer the comment is only about BeforeNodeDeletedEvent, and to group by listening class.

come-nc commented on apps/files_trashbin/lib/Storage.php:
```diff
@@ -37,48 +37,31 @@
+		?IUserManager $userManager = null,
+		?LoggerInterface $logger = null,
+		?EventDispatcherInterface $eventDispatcher = null,
+		?IRootFolder $rootFolder = null
```
When are those null? A lot of the code is expecting non-null values.

CarlSchwan replied:
I just adapted the old php doc types. Just to be sure to not have regressions. If we want to make them non nullable, best to do in another PR to make it easier to bisect.

come-nc replied:
Why is the constructor in the interface?

CarlSchwan replied:
No idea :(

come-nc replied:
This will be dealt with in another PR as this constructor is not touched anymore by the PR after rebase.

come-nc commented on apps/files_trashbin/lib/Trashbin.php:
```diff
@@ -44,6 +44,7 @@
  */
 namespace OCA\Files_Trashbin;
 
+use OC\Files\Node\Node;
```
```suggestion
use OC\Files\Node;
```
psalm seems to suggest the wrong Node is imported, but not sure if this is the right fix

CarlSchwan replied:
it wants the OCP type :)

tcitworld commented:
Rebased

tcitworld commented on apps/files_trashbin/lib/Trashbin.php:
```diff
@@ -63,22 +67,20 @@
 use Psr\Log\LoggerInterface;
 
-class Trashbin {
+class Trashbin implements IEventListener {
```
**question:** I don't understand why it needs this interface :thinking:

tcitworld commented on lib/composer/composer/autoload_classmap.php:
```diff
@@ -954,6 +955,7 @@
     'OC\\Core\\Command\\Maintenance\\RepairShareOwnership' => $baseDir . '/core/Command/Maintenance/RepairShareOwnership.php',
     'OC\\Core\\Command\\Maintenance\\UpdateHtaccess' => $baseDir . '/core/Command/Maintenance/UpdateHtaccess.php',
     'OC\\Core\\Command\\Maintenance\\UpdateTheme' => $baseDir . '/core/Command/Maintenance/UpdateTheme.php',
+    'OC\\Core\\Command\\Preview\\Regenerate' => $baseDir . '/core/Command/Preview/Regenerate.php',
```
**nitpick:** Don't know where this comes from

skjnldsv commented:
As there is no feedback since a while I will close this ticket. 
If you will decide to work on this feature again and if it hasn't been fixed or implemented already, feel free to re-open and solve the various conflicts.

come-nc committed:
chore: Use SPDX headers and property promotion in new files

Signed-off-by: Côme Chilliet <come.chilliet@nextcloud.com>

come-nc committed:
fix(tests): Fix tests now that trashbin listens to events properly

Hooks are cleared in test bootstrap so switching to events activates
 them in tests.

Signed-off-by: Côme Chilliet <come.chilliet@nextcloud.com>

come-nc committed:
fix(tests): Pause trashbin for lock tests in ViewTest using rmdir and unlink

Signed-off-by: Côme Chilliet <come.chilliet@nextcloud.com>

come-nc commented:
The 4 ViewTest failures came from the fact that the trashbin wrapper is now registered and intercepts the rmdir and unlink calls to applies trashbin logic instead of actually deleting stuff.
I added a call to pause trash in those tests and it passes now.
